tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -40,238 +86,263 @@ auto read_json(std::filesystem::path const& path) {
+                  << ", attribute: " << attribute_name
+                  << ": actual = " << get_as_string<decltype(attribute_value)>(attribute_value) + " vs. expected = "
+                  << get_as_string<decltype(ref_attribute_value)>(ref_attribute_value);
+        CHECK_MESSAGE(match, case_sstr.str());
     }
 }
 
-bool check_angle_and_magnitude(RawDataConstPtr reference_result_ptr, RawDataConstPtr result_ptr,
-                               MetaAttribute const& angle_attr, MetaAttribute const& mag_attr, double atol, double rtol,
-                               Idx obj) {
-    if (angle_attr.ctype == CType::c_double) {
-        assert(mag_attr.ctype == CType::c_double);
-        return check_angle_and_magnitude<symmetric_t>(reference_result_ptr, result_ptr, angle_attr, mag_attr, atol,
-                                                      rtol, obj);
+template <typename T>
+void check_individual_attribute(Buffer const& buffer, Buffer const& ref_buffer,
+                                MetaAttribute const* const attribute_meta,
+                                MetaAttribute const* const possible_attr_meta, bool const& is_angle,
+                                Idx elements_per_scenario, Idx scenario_idx, Idx obj, std::string const& component_name,
+                                std::string const& attribute_name, double const& dynamic_atol, double const& rtol) {
+    Idx idx = (elements_per_scenario * scenario_idx) + obj;
+    // get attribute values to check
+    T ref_attribute_value{};
+    T attribute_value{};
+    T ref_possible_attribute_value{};
+    T possible_attribute_value{};
+    auto get_values = [&ref_buffer, &buffer, &attribute_meta, &possible_attr_meta,
+                       idx]<typename U>(U* ref_value, U* value, U* ref_possible_value, U* possible_value) {
+        ref_buffer.get_value(attribute_meta, ref_value, idx, 0);
+        buffer.get_value(attribute_meta, value, idx, 0);
+        ref_buffer.get_value(possible_attr_meta, ref_possible_value, idx, 0);
+        buffer.get_value(possible_attr_meta, possible_value, idx, 0);
+    };
+    if constexpr (std::is_same_v<std::decay_t<T>, std::array<double, 3>>) {
+        get_values(ref_attribute_value.data(), attribute_value.data(), ref_possible_attribute_value.data(),
+                   possible_attribute_value.data());
+    } else {
+        get_values(&ref_attribute_value, &attribute_value, &ref_possible_attribute_value, &possible_attribute_value);
+    }
+
+    // check attributes
+    if (is_nan(ref_attribute_value)) {
+        return;
     }
-    assert(angle_attr.ctype == CType::c_double3);
-    assert(mag_attr.ctype == CType::c_double3);
-    return check_angle_and_magnitude<asymmetric_t>(reference_result_ptr, result_ptr, angle_attr, mag_attr, atol, rtol,
-                                                   obj);
+    // for angle attribute, also check the magnitude available
+    if (is_angle && is_nan(ref_possible_attribute_value)) {
+        return;
+    }
```

Comment:
Why is it that in a `check*` function some `nan` would only yeild `void return`s but nothing else?

Reply:
Because if the reference attributes are `nan`, nothing is to be checked, they can be skipped. This is needed because of the way the reference result dataset is created. This way we only check for the results we actually want to check.

Reply:
In that case, perhaps we could do a simple `should_skip_check` function to do these two `if` statements and only do the check if otherwise:
```
    // check attributes
    if (!should_skip_check(ref_attribute_value, is_angle, ref_possible_attribute_value)) {
        check_results(ref_attribute_value, attribute_value, ref_possible_attribute_value, possible_attribute_value,
                      is_angle, scenario_idx, obj, component_name, attribute_name, dynamic_atol, rtol);
    }
}
```

Reply:
I like this approach, a lot more readable. Resolving now, the implementation will be reflected in the following commit.